[quant][graphmode] Replicate quantize node for prim::If

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35130 [quant][graphmode] Make interpolate/upsample work again
* #35142 [quant][graphmode] SwapDeQuant support prim::If
* #35141 [quant][graphmode][refactor] getGeneralOpTensorInputIndexes -> getGeneralOpTensorInputs
* #35135 [quant][graphmode][refactor] swapDeQuant takes block as arugment
* #35077 [quant][graphmode] Fold quantized prepacking ops
* #35073 [quant] Make conv2d_prepack and linear_prepack pure
* #34855 [quant][graphmode] quantization support for prim::CallFunction
* #34346 [quant][graphmode] Add quantization support for aten::cat
* #34345 [quant][graphmode] Add prim::ListConstruct to general op handling
* #34854 [quant][graphmode] quantization support for view, transpose, contiguos
* #34806 [quant][graphmode] quantization support for aten::chunk
* #34807 [quant][graphmode] quantization support for prim::ListUnpack
* **#34804 [quant][graphmode] Replicate quantize node for prim::If**

Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

jerryzh168 committed:
[quant][graphmode] Replicate quantize node for prim::If

Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar reviewed (approved):
LGTM

supriyar commented on test/test_jit.py:
```diff
@@ -2077,6 +2077,42 @@ def forward(self, x, weight, bias):
+        m = torch.jit.script(M()).eval()
+        m = quantize_script(m, qconfig_dict, _test_only_eval_fn, [data], inplace=False)
+        # make sure patterns in both branches are fused
+        FileCheck().check_count("quantized::conv2d(", 4, exactly=True) \
```
Does the test need other checks here for `quantize_per_tensor`?

jerryzh168 replied:
we have insert observers tests for prim::If as well, so `quantize_per_tensor` is already covered there.

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Replicate quantize node for prim::If"


Summary:
We want to replicate the quantize node for return values in blocks of prim::If
in order to create the quantization patterns.

Test Plan:
python test/test_jit.py

Reviewers:
.
Subscribers:

Tasks:

Tags:

Differential Revision: [D20524453](https://our.internmc.facebook.com/intern/diff/D20524453)

[ghstack-poisoned]
